doc/alttab.1
.\" generated with Ronn-NG/v0.10.1
.\" http://github.com/apjanke/ronn-ng/tree/0.10.1
.TH "ALTTAB" "1" "May 2026" ""
.SH "NAME"
\fBalttab\fR \- the task switcher
.SH "SYNOPSIS"
\fBalttab\fR [\fB\-w\fR \fIN\fR] [\fB\-d\fR \fIN\fR] [\fB\-sc\fR \fIN\fR] [\fB\-mk\fR \fIstr\fR] [\fB\-kk\fR \fIstr\fR] [\fB\-bk\fR \fIstr\fR] [\fB\-pk\fR \fIstr\fR] [\fB\-nk\fR \fIstr\fR] [\fB\-ck\fR \fIstr\fR] [\fB\-dk\fR \fIstr\fR] [\fB\-mm\fR \fIN\fR] [\fB\-bm\fR \fIN\fR] [\fB\-t\fR \fINxM\fR] [\fB\-i\fR \fINxM\fR] [\fB\-vp\fR \fIstr\fR] [\fB\-p\fR \fIstr\fR] [\fB\-s\fR \fIN\fR] [\fB\-theme\fR \fIname\fR] [\fB\-bg\fR \fIcolor\fR] [\fB\-fg\fR \fIcolor\fR] [\fB\-frame\fR \fIcolor\fR] [\fB\-inact\fR \fIcolor\fR] [\fB\-min\fR \fIcolor\fR] [\fB\-bc\fR \fIcolor\fR] [\fB\-bw\fR \fIN\fR] [\fB\-font\fR \fIname\fR] [\fB\-vertical\fR] [\fB\-sortmin\fR] [\fB\-e\fR] [\fB\-b\fR \fIN\fR] [\fB\-ns\fR] [\fB\-v\fR|\fB\-vv\fR]
.SH "DESCRIPTION"
The task switcher designed for minimalistic window managers or standalone X11 session\.
.P
Unlike task switchers integrated in most simple WMs, \fBalttab\fR features visual interface and convenient tactile behaviour, switching windows on releasing Alt key\.
.P
Being lightweight and depending only on core X11 libs, it still features customizable look and behaviour\.
.SH "OPTIONS"
Most command line options have corresponding X resource, see doc/alttab\.ad for example\.
.TP
\fB\-w\fR \fINUMBER\fR
resource: alttab\.windowmanager
.br
default: autodetect, fallback to \fI3\fR
.IP
Control how to obtain window list and set focus\. \fINUMBER\fR must be between 0 and 3:
.IP
\fI0\fR: No WM, raw X\.
.IP
Set this value manually, as it's not autodetected\. Only top\-level viewable children of root window are listed\. Running \fB\-w 0\fR under WM is discouraged: hidden windows or windows on another desktops will not be listed; instead, some strange windows may be included\.
.IP
\fI1\fR: EWMH compatible\.
.IP
Compatible WM list is in [SEE ALSO] below\. alttab functions in xmonad (EWMH must be allowed, details in doc/wm\-setup\.md), i3, evilwm, xfwm4, metacity, jwm, openbox, fluxbox, icewm, enlightenment, blackbox, aewm, fvwm, sawfish, bspwm\. The following WMs support EWMH partially but work in \fB\-w 1\fR mode too: dwm (issue #9), matchbox (issue #34), windowmaker (issue #9), xpra (issue #37), cwm (issue #35), ctwm (issue #39), lwm (issue #40), awesome (issue #41)\.
.IP
\fI2\fR: Ratpoison\.
.IP
Alttab uses ratpoison own remote control capability\. Also, alttab registers itself in its "unmanaged list" automatically\.
.IP
\fI3\fR: Ancient WM\.
.IP
Fallback, worse than \fI0\fR\. Search for application windows at random nodes of window tree (WM is probably reparenting) and select only those with a name, including non\-viewable (WM is probably iconifying)\. May cause lag\. Tested in twm, flwm, wm2\.
.TP
\fB\-d\fR \fINUMBER\fR
resource: alttab\.desktops
.br
default: \fI0\fR
.IP
Show windows belonging to desktops (workspaces, groups)\. \fINUMBER\fR must be between 0 and 4:
.IP
\fI0\fR: Current desktop\.
.IP
This is default, although in non EWMH\-compatible WM, except ratpoison, it's effectively the same as \fI1\fR\.
.IP
\fI1\fR: All desktops\.
.IP
\fI2\fR: All desktops, excluding special (\-1) one\.
.IP
In various EWMH WM this desktop may be designed hidden or containing sticky windows\. In i3, it's "scratchpad"\.
.IP
\fI3\fR: All desktops, excluding current and special\.
.IP
\fI4\fR: Special (\-1) desktop only\.
.TP
\fB\-sc\fR \fINUMBER\fR
resource: alttab\.screens
.br
default: \fI1\fR
.IP
Show windows belonging to xrandr/xinerama screens\. \fINUMBER\fR must be 0 or 1:
.IP
\fI0\fR: Current screen\.
.IP
As defined according to \fB\-vp\fR \fIpointer\fR or \fB\-vp\fR \fIfocus\fR\.
.IP
\fI1\fR: All screens\.
.IP
This is default\.
.TP
\fB\-mk\fR \fIstr\fR
resource: alttab\.modifier\.keysym
.br
default: Alt_L (i\.e\. Left Alt)
.IP
Keysym of main modifier\. String or number accepted\. Suitable examples: Control_L, Super_L\.
.TP
\fB\-kk\fR \fIstr\fR
resource: alttab\.key\.keysym
.br
default: Tab
.IP
Keysym of main key\. For example, grave (aka tilde), Escape\.
.TP
\fB\-bk\fR \fIstr\fR
resource: alttab\.backscroll\.keysym
.br
default: any Shift, according to ShiftMask
.IP
Keysym of backward scroll key\.
.TP
\fB\-pk\fR \fIstr\fR
resource: alttab\.prevkey\.keysym
.br
default: no key assigned
.IP
Keysym of auxiliary 'previous' key\. For example, Left, H\.
.TP
\fB\-nk\fR \fIstr\fR
resource: alttab\.nextkey\.keysym
.br
default: no key assigned
.IP
Keysym of auxiliary 'next' key\. For example, Right, L\.
.TP
\fB\-ck\fR \fIstr\fR
resource: alttab\.cancelkey\.keysym
.br
default: Escape
.IP
Keysym of auxiliary 'cancel' key\.
.TP
\fB\-dk\fR \fIstr\fR
resource: alttab\.killkey\.keysym
.br
default: k
.IP
Keysym of key for killing X client to which currently selected window belongs\.
.TP
\fB\-mm\fR \fINUM\fR
resource: alttab\.modifier\.mask
.IP
OBSOLETED, guessed from \-mk instead\.
.TP
\fB\-bm\fR \fINUM\fR
resource: alttab\.backscroll\.mask
.IP
OBSOLETED, guessed from \-bk instead\.
.TP
\fB\-t\fR \fINNxMM\fR
resource: alttab\.tile\.geometry
.br
default: \fI112x128\fR
.IP
Size of a single tile (mini\-window)\. If switcher doesn't fit the viewport (\fB\-vp\fR), tile may be thinner\.
.TP
\fB\-i\fR \fINNxMM\fR
resource: alttab\.icon\.geometry
.br
default: \fI32x32\fR
.IP
Size of icons\.
.TP
\fB\-vp\fR \fIstr\fR
resource: alttab\.viewport
.br
default: \fIfocus\fR
.IP
Limit viewport for the switcher\. The switcher always has variable size and position, but tries to never break out of \fB\-vp\fR container, while \fB\-p\fR specifies position relative to this container\. Together these options allow for WM\-independent multi\-monitor support\. Possible values for \fB\-vp\fR:
.IP
\fIfocus\fR: in multihead configuration: the monitor which shows largest part of currently focused window\. If this part is shared with other monitors, then the smallest of these monitors is chosen\. In single head configuration: the geometry of default root window\.
.IP
\fIpointer\fR: the monitor which has mouse pointer, otherwise the same as \fIfocus\fR\.
.IP
The "multihead" above means more than one active XRANDR output\. They are detected at runtime, so hopefully monitors may be attached/detached without restarting alttab\.
.IP
\fItotal\fR: the geometry of default root window\.
.IP
If you specify this in tiling multihead configuration, then the switcher will be drawn relative to the entire combined screen, crossing monitors' borders\.
.IP
\fIWxH+X+Y\fR: Specific position relative to default root window\.
.IP

src/alttab.c
/*
Parsing options/resources, top-level keygrab functions and main().

Copyright 2017-2025 Alexander Kulak.
This file is part of alttab program.

alttab is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

alttab is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with alttab.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xresource.h>
#include <X11/Xft/Xft.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <signal.h>
#include "alttab.h"
#include "util.h"
#include "config.h"

// PUBLIC

Globals g;
// globals common for alttab, util and icon
Display *dpy;
int scr;
Window root;

// PRIVATE
static XrmDatabase db;

//
// help and exit
//
static void helpexit(void)
{
    msg(-1, "the task switcher, v%s\n\
Options:\n\
    -w N      window manager: 0=no, 1=ewmh-compatible, 2=ratpoison, 3=old fashion\n\
    -d N      desktop: 0=current 1=all, 2=all but special, 3=all but current, 4=special\n\
   -sc N      screen: 0=current 1=all\n\
   -kk str    keysym of main key\n\
   -mk str    keysym of main modifier\n\
   -bk str    keysym of backscroll modifier\n\
   -pk str    keysym of 'prev' key\n\
   -nk str    keysym of 'next' key\n\
   -ck str    keysym of 'cancel' key\n\
   -dk str    keysym of 'kill' key\n\
   -mm N      (obsoleted) main modifier mask\n\
   -bm N      (obsoleted) backward scroll modifier mask\n\
    -t NxM    tile geometry\n\
    -i NxM    icon geometry\n\
   -vp str    switcher viewport: focus, pointer, total, WxH+X+Y\n\
    -p str    switcher position: center, none, +X+Y\n\
    -s N      icon source: 0=X11, 1=fallback to files, 2=best size, 3=files only, 4=no\n\
-theme name   icon theme\n\
   -bg color  background color\n\
   -fg color  foreground color\n\
-frame color  active frame color\n\
-inact color  inactive frame color\n\
  -min color  minimized frame color\n\
   -bc color  extra border color\n\
   -bw N      extra border width\n\
 -font name   font name in the form xft:fontconfig_pattern\n\
 -vertical    vertical layout\n\
  -sortmin    sort minimized windows last\n\
    -e        keep switcher after keys release\n\
    -b N      bottom line: 0=no, 1=desktop, 2=user\n\
   -ns        ignore window request to skip it in taskbar\n\
  -v|-vv      verbose\n\
    -h        help\n\
See man alttab for details.\n", PACKAGE_VERSION);
    exit(0);
}

//
// initialize globals based on executable arguments and Xresources
// return 1 if success, 0 otherwise
// on fatal failure, calls die/exit
//
static int use_args_and_xrm(int *argc, char **argv)
{
// set debug level early
    g.debug = 0;
    char *errmsg;
    int ksi;
    KeyCode BC;
    unsigned int wmindex, dsindex, scindex, isrc, bindex;
    char *gtile, *gicon, *gview, *gpos;
    int x, y;
    unsigned int w, h, bw;
    int xpg;
    char *s;
    char *rm;
    char *empty = "";
    int uo;
    Atom nwm_prop, atype;
    unsigned char *nwm;
    int form;
    unsigned long remain, len;
    XrmOptionDescRec xrmTable[] = {
        {"-w", "*windowmanager", XrmoptionSepArg, NULL},
        {"-d", "*desktops", XrmoptionSepArg, NULL},
        {"-sc", "*screens", XrmoptionSepArg, NULL},
        {"-mm", "*modifier.mask", XrmoptionSepArg, NULL},
        {"-bm", "*backscroll.mask", XrmoptionSepArg, NULL},
        {"-mk", "*modifier.keysym", XrmoptionSepArg, NULL},
        {"-kk", "*key.keysym", XrmoptionSepArg, NULL},
        {"-bk", "*backscroll.keysym", XrmoptionSepArg, NULL},
        {"-pk", "*prevkey.keysym", XrmoptionSepArg, NULL},
        {"-nk", "*nextkey.keysym", XrmoptionSepArg, NULL},
        {"-ck", "*cancelkey.keysym", XrmoptionSepArg, NULL},
        {"-dk", "*killkey.keysym", XrmoptionSepArg, NULL},
        {"-t", "*tile.geometry", XrmoptionSepArg, NULL},
        {"-i", "*icon.geometry", XrmoptionSepArg, NULL},
        {"-vp", "*viewport", XrmoptionSepArg, NULL},
        {"-p", "*position", XrmoptionSepArg, NULL},
        {"-s", "*icon.source", XrmoptionSepArg, NULL},
        {"-theme", "*theme", XrmoptionSepArg, NULL},
        {"-bg", "*background", XrmoptionSepArg, NULL},
        {"-fg", "*foreground", XrmoptionSepArg, NULL},
        {"-frame", "*framecolor", XrmoptionSepArg, NULL},
        {"-inact", "*inactcolor", XrmoptionSepArg, NULL},
        {"-min", "*mincolor", XrmoptionSepArg, NULL},
        {"-bc", "*bordercolor", XrmoptionSepArg, NULL},
        {"-bw", "*borderwidth", XrmoptionSepArg, NULL},
        {"-font", "*font", XrmoptionSepArg, NULL},
        {"-vertical", "*vertical", XrmoptionIsArg, NULL},
        {"-e", "*keep", XrmoptionIsArg, NULL},
        {"-b", "*bottomline", XrmoptionSepArg, NULL},
        {"-sortmin", "*sortmin", XrmoptionIsArg, NULL},
        {"-ns", "*noskiptaskbar", XrmoptionIsArg, NULL}
    };
    const char *inv = "invalid %s, use -h for help\n";
    const char *rmb = "can't figure out modmask from keycode 0x%x\n";

// not using getopt() because of need for "-v" before Xrm
    int arg;
    for (arg = 0; arg < (*argc); arg++) {
        if ((strcmp(argv[arg], "-v") == 0)) {
            g.debug = 1;
            remove_arg(argc, argv, arg);
        } else if ((strcmp(argv[arg], "-vv") == 0)) {
            g.debug = 2;
            remove_arg(argc, argv, arg);
        } else if ((strcmp(argv[arg], "-h") == 0)) {
            helpexit();
            remove_arg(argc, argv, arg);
        }
    }
    msg(0, "%s\n", PACKAGE_STRING);
    msg(0, "debug level %d\n", g.debug);

    XrmInitialize();
    rm = XResourceManagerString(dpy);
    msg(1, "resource manager: \"%s\"\n", rm);
    if (!rm) {
        msg(0, "can't get resource manager, using empty db\n");
        //return 0;  // we can do it
        //db = XrmGetDatabase (dpy);
        rm = empty;
    }
    db = XrmGetStringDatabase(rm);
    if (!db) {
        msg(-1, "can't get resource database\n");
        return 0;
    }
    XrmParseCommand(&db, xrmTable, sizeof(xrmTable) / sizeof(xrmTable[0]),
                    XRMAPPNAME, argc, argv);
    if ((*argc) > 1) {
        g.debug = 1;
        msg(-1, "unknown options or wrong arguments:");
        for (uo = 1; uo < (*argc); uo++) {
            msg(0, " \"%s\"", argv[uo]);
        }
        msg(0, ", use -h for help\n");
        exit(1);
    }

    switch (xresource_load_int(&db, XRMAPPNAME, "windowmanager", &wmindex)) {
    case 1:
        if (wmindex >= WM_MIN && wmindex <= WM_MAX) {
            g.option_wm = wmindex;
            goto wmDone;
        } else {
            die(inv, "windowmanager argument range");
        }
        break;
    case 0:
        msg(0, "no WM index or unknown, guessing\n");
        break;
    case -1:
        die(inv, "windowmanager argument");
        break;
    }
// EWMH?
    if (ewmh_detectFeatures(&(g.ewmh))) {
        msg(0, "EWMH-compatible WM detected: %s\n", g.ewmh.wmname);
        g.option_wm = WM_EWMH;
        goto wmDone;
    }
// ratpoison?
    nwm_prop = XInternAtom(dpy, "_NET_WM_NAME", false);
    if (XGetWindowProperty(dpy, root, nwm_prop, 0, MAXNAMESZ, false,
                           AnyPropertyType, &atype, &form, &len, &remain,
                           &nwm) == Success && nwm) {
        msg(0, "_NET_WM_NAME root property present: %s\n", nwm);
        if (strstr((char *)nwm, "ratpoison") != NULL) {
            g.option_wm = WM_RATPOISON;
            XFree(nwm);
            goto wmDone;
        }
        XFree(nwm);
    }
    msg(0, "unknown WM, using WM_TWM\n");
    g.option_wm = WM_TWM;
 wmDone:
    msg(0, "WM: %d\n", g.option_wm);

    switch (xresource_load_int(&db, XRMAPPNAME, "desktops", &dsindex)) {
    case 1:
        if (dsindex >= DESK_MIN && dsindex <= DESK_MAX)
            g.option_desktop = dsindex;
        else
            die(inv, "desktops argument range");
        break;
    case 0:
        g.option_desktop = DESK_DEFAULT;
        break;
    case -1:
        die(inv, "desktops argument");
        break;
    }
    msg(0, "desktops: %d\n", g.option_desktop);

    switch (xresource_load_int(&db, XRMAPPNAME, "screens", &scindex)) {
    case 1:
        if (scindex >= SCR_MIN && scindex <= SCR_MAX)
            g.option_screen = scindex;
        else
            die(inv, "screens argument range");
        break;
    case 0:
        g.option_screen = SCR_DEFAULT;
        break;
    case -1:
        die(inv, "screens argument");
        break;
    }
    msg(0, "screens: %d\n", g.option_screen);

#define  MC  g.option_modCode
#define  KC  g.option_keyCode
#define  prevC  g.option_prevCode
#define  nextC  g.option_nextCode
#define  cancelC  g.option_cancelCode
#define  killC  g.option_killCode
#define  GMM  g.option_modMask
#define  GBM  g.option_backMask

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "modifier", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    MC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFMODKS);

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "key", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    KC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFKEYKS);

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "prevkey", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    prevC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFPREVKEYKS);

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "nextkey", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    nextC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFNEXTKEYKS);

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "cancelkey", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    cancelC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFCANCELKS);

    ksi = ksym_option_to_keycode(&db, XRMAPPNAME, "killkey", &errmsg);
    if (ksi == -1)
        die("%s\n", errmsg);
    killC = ksi != 0 ? ksi : XKeysymToKeycode(dpy, DEFKILLKS);

    switch (xresource_load_int(&db, XRMAPPNAME, "modifier.mask", &(GMM))) {
    case 1:
        msg(-1,
            "Using obsoleted -mm option or modifier.mask resource, see man page for upgrade\n");
        break;
    case 0:
        GMM = keycode_to_modmask(MC);
        if (GMM == 0)
            die(rmb, MC);
        break;
    case -1:
        die(inv, "modifier mask");
        break;
    }

    switch (xresource_load_int(&db, XRMAPPNAME, "backscroll.mask", &(GBM))) {
    case 1:
        msg(-1,
            "Using obsoleted -bm option or backscroll.mask resource, see man page for upgrade\n");
        break;
    case 0:
        BC = ksym_option_to_keycode(&db, XRMAPPNAME, "backscroll", &errmsg);
        if (BC != 0) {
            GBM = keycode_to_modmask(BC);
            if (GBM == 0)
                die(rmb, BC);
        } else {
            GBM = DEFBACKMASK;
        }
        break;
    case -1:
        die(inv, "backscroll mask");
        break;
    }

    msg(0, "modMask %d, backMask %d, modCode %d, keyCode %d\n",
        GMM, GBM, MC, KC);
    msg(0, "cancelCode %d, killCode %d\n",
        cancelC, killC);

    g.option_tileW = DEFTILEW;
    g.option_tileH = DEFTILEH;
    gtile = xresource_load_string(&db, XRMAPPNAME, "tile.geometry");
    if (gtile != NULL) {
        xpg = XParseGeometry(gtile, &x, &y, &w, &h);
        if (xpg & WidthValue)
            g.option_tileW = w;
        else
            die(inv, "tile width");
        if (xpg & HeightValue)
            g.option_tileH = h;
        else
            die(inv, "tile height");
    }
    if (g.option_tileW == 0 || g.option_tileH == 0)
        die(inv, "tile width or height: can't be zero");

    g.option_iconW = DEFICONW;
    g.option_iconH = DEFICONH;
    gicon = xresource_load_string(&db, XRMAPPNAME, "icon.geometry");
    if (gicon) {
        xpg = XParseGeometry(gicon, &x, &y, &w, &h);
        if (xpg & WidthValue)
            g.option_iconW = w;
        else
            die(inv, "icon width");
        if (xpg & HeightValue)
            g.option_iconH = h;
        else
            die(inv, "icon height");
    }
    if (g.option_iconW == 0 || g.option_iconH == 0)
        die(inv, "icon width or height: can't be zero");

    msg(0, "%dx%d tile, %dx%d icon\n",
        g.option_tileW, g.option_tileH, g.option_iconW, g.option_iconH);

    switch (xresource_load_int(&db, XRMAPPNAME, "borderwidth", &bw)) {
    case 1:
        if (bw >= BORDER_MIN)
          g.option_borderW = bw;
        else
          die(inv, "bw argument range");
        break;
    case 0:
        g.option_borderW = DEFBORDERW;
        break;
    case -1:
        die(inv, "bw argument");
        break;
    }
    msg(0, "bw: %d\n", g.option_borderW);

    bzero(&(g.option_vp), sizeof(g.option_vp));
    g.option_vp_mode = VP_DEFAULT;
    gview = xresource_load_string(&db, XRMAPPNAME, "viewport");
    if (gview) {
        if (strncmp(gview, "focus", 6) == 0) {
            g.option_vp_mode = VP_FOCUS;
        } else if (strncmp(gview, "pointer", 8) == 0) {
            g.option_vp_mode = VP_POINTER;
        } else if (strncmp(gview, "total", 6) == 0) {
            g.option_vp_mode = VP_TOTAL;
        } else {
            g.option_vp_mode = VP_SPECIFIC;
            xpg = XParseGeometry(gview, &x, &y, &w, &h);
            if (xpg & (XValue | YValue | WidthValue | HeightValue)) {
                g.option_vp.w = w;
                g.option_vp.h = h;
                g.option_vp.x = x;
                g.option_vp.y = y;
            } else {
                die(inv, "viewport");
            }
        }
    }
    msg(0, "viewport: mode %d, %dx%d+%d+%d\n",
        g.option_vp_mode,
        g.option_vp.w, g.option_vp.h, g.option_vp.x, g.option_vp.y);

    g.option_positioning = POS_DEFAULT;
    g.option_posX = 0;
    g.option_posY = 0;
    gpos = xresource_load_string(&db, XRMAPPNAME, "position");
    if (gpos) {
        if (strncmp(gpos, "center", 7) == 0) {
            g.option_positioning = POS_CENTER;
        } else if (strncmp(gpos, "none", 5) == 0) {
            g.option_positioning = POS_NONE;
        } else {
            g.option_positioning = POS_SPECIFIC;
            xpg = XParseGeometry(gpos, &x, &y, &w, &h);
            if (xpg & (XValue | YValue)) {
                g.option_posX = x;
                g.option_posY = y;
            } else {
                die(inv, "position");
            }
        }
    }
    msg(0, "positioning policy: %d, position: +%d+%d\n",
        g.option_positioning, g.option_posX, g.option_posY);

    g.option_iconSrc = ISRC_DEFAULT;
    switch (xresource_load_int(&db, XRMAPPNAME, "icon.source", &isrc)) {
    case 1:
        if (isrc >= ISRC_MIN && isrc <= ISRC_MAX)
            g.option_iconSrc = isrc;
        else
            die("icon source argument must be from %d to %d\n",
                ISRC_MIN, ISRC_MAX);
        break;
    case 0:
        g.option_iconSrc = ISRC_DEFAULT;
        break;
    case -1:
        die(inv, "icon source");
        break;
    }
    msg(0, "icon source: %d\n", g.option_iconSrc);

    s = xresource_load_string(&db, XRMAPPNAME, "theme");
    g.option_theme = s ? s : DEFTHEME;
    msg(0, "icon theme: %s\n", g.option_theme);

    s = xresource_load_string(&db, XRMAPPNAME, "background");
    g.color[COLBG].name = s ? s : DEFCOLBG;
    s = xresource_load_string(&db, XRMAPPNAME, "foreground");
    g.color[COLFG].name = s ? s : DEFCOLFG;
    s = xresource_load_string(&db, XRMAPPNAME, "framecolor");
    g.color[COLFRAME].name = s ? s : DEFCOLFRAME;
    
    // workaround to match COLMIN to COLFRAME w/o new variables
    g.color[COLMIN].name = s ? s : DEFCOLFRAME; // Set minimized color to frame color
    s = xresource_load_string(&db, XRMAPPNAME, "mincolor");
    if (s) { // if minimized color was specified, overwrite
        g.color[COLMIN].name = s;
        g.option_minimize_color = (s != NULL);
    }

    s = xresource_load_string(&db, XRMAPPNAME, "inactcolor");
    g.color[COLINACT].name = s ? s : g.color[COLFG].name;
    s = xresource_load_string(&db, XRMAPPNAME, "bordercolor");
    g.color[COLBORDER].name = s ? s : DEFCOLBORDER;

    s = xresource_load_string(&db, XRMAPPNAME, "font");
    if (s) {
        if ((strncmp(s, "xft:", 4) == 0)
            && (*(s + 4) != '\0')) {
            g.option_font = s + 4;
        } else {
            // resource may indeed be valid but non-xft
            msg(-1, "invalid font: %s, using default: %s\n", s, DEFFONT);
            g.option_font = DEFFONT + 4;
        }
    } else {
        g.option_font = DEFFONT + 4;
    }

    s = xresource_load_string(&db, XRMAPPNAME, "vertical");
    g.option_vertical = (s != NULL);
    msg(0, "vertical: %d\n", g.option_vertical);

    s = xresource_load_string(&db, XRMAPPNAME, "sortmin");
    g.option_sort_minimize = (s != NULL);
    msg(0, "sortmin: %d\n", g.option_sort_minimize);

// max recursion for searching windows
// -1 is "everything"
// in raw X this returns too much windows, "1" is probably sufficient
// no need for an option
    g.option_max_reclevel = (g.option_wm == WM_NO) ? 1 : -1;

    s = xresource_load_string(&db, XRMAPPNAME, "keep");
    g.option_keep_ui = (s != NULL);
    msg(0, "keep_ui: %d\n", g.option_keep_ui);

    switch (xresource_load_int(&db, XRMAPPNAME, "bottomline", &bindex)) {
    case 1:
        if (bindex >= BL_MIN && bindex <= BL_MAX)
            g.option_bottom_line = bindex;
        else
            die(inv, "bottomline argument range");
        break;
    case 0:
        g.option_bottom_line = BL_NONE;
        break;
    case -1:
        die(inv, "bottomline argument");
        break;
    }
    msg(0, "bottomline: %d\n", g.option_bottom_line);

    s = xresource_load_string(&db, XRMAPPNAME, "noskiptaskbar");
    g.option_no_skip_taskbar = (s != NULL);
    msg(0, "no_skip_taskbar: %d\n", g.option_no_skip_taskbar);

    return 1;
}

//
// grab Alt-Tab and Alt-Shift-Tab
// note: exit() on failure
//
static int grabKeysAtStartup(bool grabUngrab)
{
    g.ignored_modmask = getOffendingModifiersMask(); // or 0 for g.debug
    char *grabhint =
        "Error while (un)grabbing key 0x%x with mask 0x%x/0x%x.\nProbably other program already grabbed this combination.\nCheck: xdotool keydown alt+Tab; xdotool key XF86LogGrabInfo; xdotool keyup Tab; sleep 1; xdotool keyup alt\nand then look for active device grabs in /var/log/Xorg.0.log\nOr try Ctrl-Tab instead of Alt-Tab:  alttab -mk Control_L\n";
// attempt XF86Ungrab? probably too invasive
    if (!changeKeygrab
        (root, grabUngrab, g.option_keyCode, g.option_modMask,
         g.ignored_modmask)) {
        die(grabhint, g.option_keyCode, g.option_modMask, g.ignored_modmask);
    }
    if (!changeKeygrab
        (root, grabUngrab, g.option_keyCode,
         g.option_modMask | g.option_backMask, g.ignored_modmask)) {
        die(grabhint, g.option_keyCode,
            g.option_modMask | g.option_backMask, g.ignored_modmask);
    }

    return 1;
}

//
// Returns 0 if not an extra prev/next keycode, 1 if extra prev keycode, and 2 if extra next keycode.
//
static int isPrevNextKey(unsigned int keycode)
{
    if (keycode == g.option_prevCode) {
        return 1;
    }
    if (keycode == g.option_nextCode) {
        return 2;
    }
    // if here then is neither
    return 0;
}

// see #97
#define CHECK_97 \
    XQueryKeymap(dpy, keys_pressed); \
    if (!(keys_pressed[octet] & kmask)) { \
        msg(1, "Wrong modifier, skip event\n"); \
        continue; \
    }

int main(int argc, char **argv)
{

    XEvent ev;
    dpy = XOpenDisplay(NULL);
    if (!dpy)
        die("can't open display");
    scr = DefaultScreen(dpy);
    root = DefaultRootWindow(dpy);

    ee_complain = true;
    //hnd = (XErrorHandler)0;
    XErrorHandler hnd = XSetErrorHandler(zeroErrorHandler); // for entire program
    if (hnd) ;;                 // make -Wunused happy

    signal(SIGUSR1, sighandler);

    if (!use_args_and_xrm(&argc, argv))
        die("use_args_and_xrm failed");
    if (!startupWintasks())
        die("startupWintasks failed");
    if (!startupGUItasks())
        die("startupGUItasks failed");

    grabKeysAtStartup(true);
    g.uiShowHasRun = false;

    struct timespec nanots;
    nanots.tv_sec = 0;
    nanots.tv_nsec = 1E7;
    char keys_pressed[32];
    int octet = g.option_modCode / 8;
    int kmask = 1 << (g.option_modCode - octet * 8);

    while (true) {
        memset(&(ev.xkey), 0, sizeof(ev.xkey));

        if (g.uiShowHasRun && ! g.option_keep_ui) {
            // poll: lag and consume cpu, but necessary because of bug #1 and #2
            XQueryKeymap(dpy, keys_pressed);
            if (!(keys_pressed[octet] & kmask)) {   // Alt released
                uiHide();
                continue;
            }
            if (!XCheckIfEvent(dpy, &ev, *predproc_true, NULL)) {
                nanosleep(&nanots, NULL);
                continue;
            }
        } else {
            // event: immediate, when we don't care about Alt release
            XNextEvent(dpy, &ev);
        }

        switch (ev.type) {
        case KeyPress:
            msg(1, "Press %lx: %d-%d\n",
                ev.xkey.window, ev.xkey.state, ev.xkey.keycode);
            if (ev.xkey.state & g.option_modMask) {  // alt
                if (ev.xkey.keycode == g.option_keyCode) {  // tab
                    CHECK_97;
                    if (!g.uiShowHasRun) {
                        uiShow((ev.xkey.state & g.option_backMask));
                    } else {
                        if (ev.xkey.state & g.option_backMask) {
                            uiPrevWindow();
                        } else {
                            uiNextWindow();
                        }
                    }
                } else if (ev.xkey.keycode == g.option_cancelCode) { // escape
                    CHECK_97;
                    uiSelectWindow(0);
                    uiHide();
                } else if (ev.xkey.keycode == g.option_killCode) { // k
                    CHECK_97;
                    uiKillWindow();
                } else {  // non-tab
                    switch (isPrevNextKey(ev.xkey.keycode)) {
                    case 1:
                        uiPrevWindow();
                        break;
                    case 2:
                        uiNextWindow();
                        break;
                    }
                }
            }
            break;

        case KeyRelease:
            msg(1, "Release %lx: %d-%d\n",
                ev.xkey.window, ev.xkey.state, ev.xkey.keycode);
            // interested only in "final" release
            if (!((ev.xkey.state & g.option_modMask)
                  && ev.xkey.keycode == g.option_modCode && g.uiShowHasRun)) {
                break;
            }
            if (g.option_keep_ui) {
                break;
            }
            uiHide();
            break;

        case Expose:
            if (g.uiShowHasRun) {
                uiExpose();
            }
            break;

        case ButtonPress:
        case ButtonRelease:
            uiButtonEvent(ev.xbutton);
            break;

        case PropertyNotify:
            winPropChangeEvent(ev.xproperty);
            break;

        case DestroyNotify:
            winDestroyEvent(ev.xdestroywindow);
            break;

        case FocusIn:
            winFocusChangeEvent(ev.xfocus);
            break;

        default:
            msg(1, "Event type %d\n", ev.type);
            break;
        }

    }

// this is probably never reached
    shutdownWin();
    shutdownGUI();
    XrmDestroyDatabase(db);
    grabKeysAtStartup(false);
// not restoring error handler
    XCloseDisplay(dpy);
    return 0;
} // main
